### ISSUE TITLE
Is it possible to create a TEE-agnostic attestation API for applications to use?

### ISSUE THREAD

**Issue by jhand2** [openenclave/openenclave#2589]
Currently attestation APIs are primarily SGX-specific. For example, oe_get_quote is an SGX-only API. There are many other APIs (host and enclave side) which are SGX only. Is it possible to come up with something more generic that wouldn't require an enclave application to know what TEE they're running in?

**Comment by dthaler** [comment]
There is no oe_get_quote in any public header file I see, the basic API should already be TEE-agnostic.  Where are you seeing oe_get_quote?

**Comment by yentsanglee** [comment]
Plugin APIs are designed for TEE-agnostic attestation. The first parameter of oe_get_evidence() is the evidence format id which can be something like OE_SGX_LOCAL, OE_SGX_REMOTE or OE_TEE_XYZ. The output of oe_get_evidence() can be verified by oe_verify_evidence().

**Comment by bodzhang** [comment]
The TEE attester/verifier plugin [design doc](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/CustomAttestation.md)  proposes a TEE-agnostic API to get the Attestation evidence:

    oe_result_t (*get_evidence)(
        oe_attester_t* context,
        uint32_t flags,
        const oe_claim_t* custom_claims,
        size_t custom_claims_length,
        const void* opt_params,
        size_t opt_params_size,
        uint8_t** evidence_buffer,
        size_t* evidence_buffer_size,
        uint8_t** endorsements_buffer,
        size_t* endorsements_buffer_size);

A different issue discussion already points out flags parameter might be SGX specific.  In my opinion, the opt_params support, defined as "optional plugin-specific input parameters" can also break the TEE-agnostic property we are after. If a plugin requires the opt-params, TEE-agnostic enclave code which leaves opt_params as NULL, will fail the call. The requirement should be that every plug-in must support opt_params=NULL, but can support plugin-specific op_params for non-TEE-agnostic callers.

Another issue is on custom_claims parameter. The doc doe not specify how the plugin should handle the data. For a SGX plugin, the custom_claims will be hashed and the hash is included in the signed evidence. But if plug-in is free to decide how to process the custom_claims parameter, for example, a non-SGX plugin decides to include the custom_claims as plaintext in the signed evidence, a TEE-agnostic protocol utilizing the remote attestation capability would have to deal with different security property the signed evidence is providing. A plugin that only exposes the hash of the customer_claims in the signed evidence allows the protocol designer to bind a shared secret with the signed evidence (In DH protocol or SIGMA protocol, the two peers can establishing shared secret based on public key exchange, and might want to bind the shared secret with the attestation evidence), while another plugin that expose the customer_claims plaintext in the signed evidence won't be able to support customer_claims with shared secret.  I propose that we specify the following for customer_claims processing:
1. The plugin must hash customer_claims and include the hash in the signed evidence  - Might also need to specify the exact SHA algorithm
2. The plugin shall not include the customer_claims plain text in the signed evidence. The caller can append the customer_claims plain text when assembling the protocol message.

**Comment by dthaler** [comment]
> For opt_params, the intention is for it to be provided by a verifier. 
> When a verifier sends to an attester a list of acceptable format IDs,...

Why should it be provided by a Verifier?   In the IETF attestation standardization discussions I've been in, the evidence format is up to the attester, not the verifier which might receive evidence in different formats from different attesters.  See https://tools.ietf.org/html/draft-ietf-rats-architecture

**Comment by dthaler** [comment]
The attestation architecture needs to scale down to attesters that are MCU's and so requiring them to support multiple formats is not really feasible, hence the architecture I mentioned.

**Comment by shnwc** [comment]
The attestation design has no requirement for an attester to support more than one format, and only some formats require opt_params data in order to generate evidence that are accepted by a verifier.

I  see the list of format IDs (along with their opt_params as needed) from a verifier as a request to an attester, so that the attester knows what evidence formats are acceptable by the verifier. This part of the flow is not in the scope of the IETF standard. Eventually, It's up to the attester as to which specific format evidence it generates, and it's up to the verifier to decide whether it accepts an evidence and place trust to the attester that produced the evidence.

**Comment by bodzhang** [comment]
What's the existing usage of opt_params for SGX attestation and TrustZone attestation? I'm wondering whether a TEE-specific opt_params can really be made transparent to the enclave code that needs to generate the Attestation data , and only the verifer needs to understand opt_params.

**Comment by yentsanglee** [comment]
opt_params is as an optional parameter that verifier plugin provides and attester plugin consumes when generating evidence. It becomes part of the evidence for the verifier plugin to validate. It is transparent to the enclave code and the host code.

**Comment by bodzhang** [comment]
@yentsanglee, do you have some examples of opt_params provided by the SGX verifier plugin and TrustZone verifier plugin? My concern is that the restriction of opt_params is transparent to the enclave/host code that invokes the attester plugin significantly limit the usefulness of opt_params.

**Comment by bodzhang** [comment]
[https://github.com/openenclave/openenclave/issues/2571](https://github.com/openenclave/openenclave/issues/2571) discusses OE_SGX_LOCAL_ATTESTATION_UUID as an evidence format, in parallel with OE_SGX_ECDSA_P256_ATTESTATION_UUID. For local attestation, the verifier will be running in the enclave app too. So for local attestation, the concept that the evidence format details and evidence format specific op_params can be transparent to the attester, and only exposed to the verifier, does not apply.

I think the concept of local attestation and remote attestation apply to all TEEs. A TrustZone TEE or a VSM TEE can support two enclaves inside the trusted OS communicating with each other, potentially with an attestation/verification scheme different from the case where the verifier is not within the trusted OS. Anyone different opinion about local attestation concept's applicability to non-SGX TEEs?

**Comment by dthaler** [comment]
Local and remote attestation are just special cases in my view of a more general multiple-formats-and-multiple-verifiers architecture.  More generally there can be multiple verifiers (regardless of whether one is local or just multiple remote ones).  So I think "local attestation" just means "attestation using a format recognized by a particular verifier, which happens to be local".

Certainly an OP-TEE/TZ mechanism can support two enclaves attesting to each other locally, but it doesn't require a separate format to do so.   (One could have a separate format, just like two remote verifiers could have separate formats.)

**Comment by shnwc** [comment]
In an effort to try to clarify local vs remote attestation, claims vs optional parameters, attester – verifier security model, TEE Agnostic design, etc., I put together a document with [Notes on OE SDK Attestation API and Plugin Library Design](https://github.com/shnwc/openenclave/blob/master/docs/DesignDocs/NotesOnAttestationAPI.md). I'll see if / when it's worthwhile to create an PR to push it upstream.

@bodzhang for SGX local attestation, the opt_params is the verifier enclave's SGX target info. This blob is transparent to the application-level enclave code in the attester, the OE SDK implementation for API `oe_get_evidence()` treats the opt_params as the SGX target info against which to generate an SGX report as the evidence. So only the verifier that supplied the SGX target info can verify this evidence.

For SGX EPID evidence, the opt_param is the service-provider ID (SPID) that the verifier supplies to the attester. Again, the attester application-level enclave code does not parse or interpret this data. The SGX EPID attester plugin underneath API `oe_get_evidence()` includes this SPID in its EPID quote as part of the evidence. So the EPID evifier that supplied the SPID can verify that it is covered by the evidence.

@dthaler I think in my notes document, the description in the "Attester – Verifier Security Model and TEE Agnostic Design" section aligns with your comments on the local vs remote attestation.

**Comment by yentsanglee** [comment]
TEE-agnostic attestation API has been discussed and tracked by the following issues and PRs:

- [Issue 2949](https://github.com/openenclave/openenclave/issues/2949)
- [PR 2801](https://github.com/openenclave/openenclave/issues/2801)

Per discussion in SIG-Attestation meeting on 6/11/2020, closing the issue.

**Comment by dthaler** [comment]
Other related PRs:

- PR #3089
-  PR #2959
